### ISSUE TITLE
why there is no sgd-style implementation？

### ISSUE THREAD

**Issue by brisker** [sail-sg/Adan#14]
why there is no sgd-style implementation or experiments？
（lemma 1 in the paper)

**Comment by XingyuXie** [comment]
Hi, @brisker,
The reformulated AGD in Lemma 1 is equivalent to the original AGD (GD with Nesterov momentum). In fact, most SGD implementations, e.g., that in PyTorch,  support the Nesterov-type momentum.

In the experiments, for fairness, we use the timm codebase to compare baselines, which is one of the most widely used codebases to train deep models. To train ResNet50/101, we directly adopt the SGD in timm, which uses Nesterov momentum in default for acceleration.

**Comment by brisker** [comment]
@XingyuXie 
I read the pytorch source code [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L241), but  the implementation there seems different from Lemma 1 in your paper. I can not see the (g_k - g_k-1) formula there, but it is in Lemma 1.

**Comment by XingyuXie** [comment]
If two algorithms are equivalent, especially if their outputs are the same after convergence, we do not think they will behave very differently in practice. Therefore, we naturally believe that the performance of various AGDs (including SGD with Nesterov momentum) should not differ too much.

If you are interested in the reformulated AGD, we may test its performance on ViT. We have not tried it in practice before, and it is only the intermediate product of derivation.

**Comment by brisker** [comment]
Back to Adan，why do you add the （g_k-g_k-1) item？ **I cannot find specific explaination or motivation in the paper**，but in my view，this item seems to be the crucial point of Adan. So I am a little confused by this. （Bsides，why there is only Adan，but no sgd-Adan，which is already discussed above，but still，I wonder if that item also works for sgd.)  @XingyuXie

**Comment by XingyuXie** [comment]
How to get grad difference $(g_k-g_{k-1})$ is shown in Lemma 1. It is not a heuristic term, and it comes from the reformulation of the classical AGD. You may treat the grad difference as a correction term like the extrapolation in classical AGD or wish it provides higher-order information since the grad difference connects to the local hessian. Bzw, we cannot name the reformulated AGD as sgd-Adan since it is indeed the AGD.

If we remove the grad difference from the reformulated AGD in Lemma 1, it will degenerate to the classical heavy-ball acceleration (HBA). Hence, your question may become *Whether AGD is better than HBA*?. This is a question that has been discussed for a long time. In theory, AGD seems better in several cases. But in practice, it is hard to choose the winner. For DNN training,  whether to enable Nesterov momentum in SGD depends on the situation. The default choice in Pytorch is False; however, Timm chooses to open it. Bzw, Nesterov momentum is seldom mentioned for the adaptive optimizers.

**Comment by brisker** [comment]
1. So， the main contribution of the paper may be ，to include the gradient difference item into adaptive optimizer， which convert Adam into Adan？
2. in the Nestro implementation  code [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L241) ，I can not see the specific grad difference item ，I am a little confused. Is that an approximation？
 @XingyuXie

**Comment by XingyuXie** [comment]
1. the contribution of Adan may refer to the paper. The comparison between Adan and Adam from theory to practice can be found there.
2. The grad difference comes from AGD. AGD can be implemented and rewritten in a variety of ways. Not every implementation has the explicit calculation for grad difference—Bzw, how to get the update formula of Nesterov momentum in your link (https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L241). Please refer to `On the importance of initialization and momentum in deep learning`.
3. A large learning rate has many advantages, such as improving generalization and accelerating convergence. A heuristic reason that allows Adan to use a large learning rate is the introduced grad diff in the second-order moment term n_k, which can slow down the update when the grad is not too consistent and accelerate the update when the grad direction is clear.
4. For more questions about the derivation of Adan rather than code, you may add me by WeChat (xyxie_joy). We may make things more clear by showing more mathematical details.

**Comment by XingyuXie** [comment]
Hi, @brisker 
I found that the [Nesterov implementation](https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L241) you mentioned is only equivalent to the original Nesterov momentum method when the objective is quadratic. Otherwise, their implementation is more like classical momentum (a.k.a. heavy-ball method).
